src/xunit.performance.api/Profilers/Etw/SimpleTraceEventParser.cs

```diff
@@ -0,0 +1,378 @@
+                        };
+                        method.LifeSpan.Start = obj.TimeStamp;
+                        method.LifeSpan.End = DateTime.MaxValue;
+
+                        module.Methods.Add(method);
+                    }
+                };
+                parser.Source.Clr.MethodUnloadVerbose += (MethodLoadUnloadVerboseTraceData obj) => {
+                    var process = processes.SingleOrDefault(p => p.Id == obj.ProcessID);
+                    if (process == null)
+                        return;
+
+                    var module = process.Modules
+                        .OfType<DotNetModule>()
+                        .SingleOrDefault(m => m.Id == obj.ModuleID);
+                    if (module == null)
+                        return;
+
+                    var method = module.Methods
+                        .SingleOrDefault(m => m.Id == obj.MethodID);
+                    if (method == null)
+                        return;
+
+                    method.LifeSpan.End = obj.TimeStamp;
+                };
+
+                source.Process();
+
+                // Map PMC to managed modules.
+                pmcRollovers
+                    .ForEach(pmc => {
+                        var performanceMonitorCounter = scenarioInfo.PerformanceMonitorCounters
+                            .SingleOrDefault(p => p.Id == pmc.ProfileSourceId);
+                        if (performanceMonitorCounter == null)
+                            return;
+
+                        processes
+                            .Single(p => p.Id == pmc.ProcessId).Modules
+                            .OfType<DotNetModule>()
+                            .ForEach(module => {
+                                var isInModule = module.AddressSpace != null
+                                    && module.LifeSpan.IsInInterval(pmc.TimeStamp)
+                                    && module.AddressSpace.IsInInterval(pmc.InstructionPointer);
+                                if (isInModule)
+                                {
+                                    if (!module.PerformanceMonitorCounterData.ContainsKey(performanceMonitorCounter))
+                                        module.PerformanceMonitorCounterData.Add(performanceMonitorCounter, 0);
+                                    module.PerformanceMonitorCounterData[performanceMonitorCounter] += pmc.SamplingInterval;
+                                    pmcRollovers.Remove(pmc);
+                                    return;
+                                }
+
+                                var methods = module.Methods
+                                    .Where(m => {
+                                        return m.LifeSpan.IsInInterval(pmc.TimeStamp)
+                                            && m.AddressSpace.IsInInterval(pmc.InstructionPointer);
+                                    })
+                                    .Select(m => m);
+                                if (methods.Count() != 0)
```

Comment:
There can potentially be hundreds of modules, thousands of methods in a module, and thousands of PMC events, so I worry about efficiency here. Do you have any measurements of how long this processing takes?

By doing this resolution in-stream instead of deferring it you don't need to worry about the time span aspect. And since method ranges generally don't overlap (eg we expect count to be either 0 or 1, and at any rate we don't distinguish between 1 and > 1), when you do this in-stream the whole things is amenable to fast lookup, say by keeping method ranges sorted by starting address.

Reply:
Currently, it takes ~7 sec to process a 280MB etl file (MusicStore.dll on CoreClr).

---
In reply to: [145202704](https://github.com/Microsoft/xunit-performance/pull/242#discussion_r145202704) [](ancestors = 145202704)

Reply:
That seems reasonably fast. W can defer working on any improvements here until we see examples that are too slow.